assemblyline_v4_service/updater/helper.py

```diff
@@ -1,26 +1,43 @@
-import certifi
 import os
-import psutil
-import regex as re
-import requests
 import shutil
 import tempfile
 import time
+import traceback
+from io import StringIO
+from logging import Logger
+from shutil import make_archive
+from typing import Any, Dict, List, Optional, Tuple
+from urllib.parse import urlparse
 
+import certifi
+import psutil
+import regex as re
+import requests
 from git import Repo
-from typing import List, Dict, Any, Tuple
-from urllib.parse import urlparse
-from shutil import make_archive
 
-from assemblyline.common.isotime import iso_to_epoch
 from assemblyline.common.digests import get_sha256_for_file
 from assemblyline.common.identify import Identify
-
+from assemblyline.common.isotime import iso_to_epoch
 
 BLOCK_SIZE = 64 * 1024
 GIT_ALLOW_UNSAFE_PROTOCOLS = os.environ.get('GIT_ALLOW_UNSAFE_PROTOCOLS', 'false').lower() == 'true'
 
-identify = Identify()
+
+DEVELOPMENT_MODE = False
+
+with StringIO() as stack_trace:
+    # Check if run_service_once, pytest or assemblyline_v4_service.testing.helper is in the stack trace to determine if we're running the service in a development mode
+    traceback.print_stack(file=stack_trace)
+    stack_trace.seek(0)
+    read_stack_trace = stack_trace.read()
+    # print(read_stack_trace)
+    if any(msg in read_stack_trace for msg in ['run_service_once', 'pytest', 'assemblyline_v4_service.testing.helper']):
+        DEVELOPMENT_MODE = True
```

Comment:
Should we have this in a small function so that this and api.py do not duplicate it?
I assume we could have something like "DEVELOPMENT_MODE = determine_dev_mode()" in those two files, and determine_dev_mode maybe coming out of utils.py?
We can also remove the commented-out print().

Reply:
if we are deduplicating I would suggest also importing `DEVELOPMENT_MODE` as a global from a single module that each uses. That way if some one else has some use case for development mode they can modify the global in one place.

Reply:
Right, just having a `from utils import DEVELOPMENT_MODE` sounds much better.